`<xutility>`: `basic_const_iterator<int *>` Cannot Convert to `basic_const_iterator<const int *>`

# Describe the bug
The conversion selects the following constructor:  

```cpp
template <convertible_to<_Iter> _Other>
constexpr basic_const_iterator(basic_const_iterator<_Other> _Current_)
    noexcept(is_nothrow_constructible_v<_Iter, _Other>) // strengthened
    : _Current(_STD move(_Current_._Current)) {}
```

However, it fails at `_Current_._Current`, which is a private member of `basic_const_iterator<int *>`. This member cannot be accessed by `basic_const_iterator<const int *>`, as the latter is not a `friend` of the former.  

# Command-line test case
```
C:\Temp>type repro.cpp
#include <iterator>

int main() {
  std::basic_const_iterator<int *> i;
  std::basic_const_iterator<const int *> ci = i;
}

C:\Temp>cl /EHsc /W4 /WX /std:c++latest .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.43.34808 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

repro.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.43.34808\include\xutility(2278): error C2248: 'std::basic_const_iterator<int *>::_Current': cannot access private member declared in class 'std::basic_const_iterator<int *>'
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.43.34808\include\xutility(2244): note: see declaration of 'std::basic_const_iterator<int *>::_Current'
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.43.34808\include\xutility(2242): note: see declaration of 'std::basic_const_iterator<int *>'
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.43.34808\include\xutility(2278): note: the template instantiation context (the oldest one first) is
.\repro.cpp(5): note: see reference to function template instantiation 'std::basic_const_iterator<const int *>::basic_const_iterator<int*>(std::basic_const_iterator<int *>) noexcept' being compiled
.\repro.cpp(5): note: see the first reference to 'std::basic_const_iterator<const int *>::basic_const_iterator' in 'main'
```

# Expected behavior
The conversion should be allowed.  

# STL version
```
Visual Studio Build Tools 2022
17.13.2
```

--- a/stl/inc/xutility
+++ b/stl/inc/xutility
@@ -2211,6 +2211,9 @@ struct _Basic_const_iterator_category<_Iter> {
 _EXPORT_STD template <input_iterator _Iter>
 class basic_const_iterator : public _Basic_const_iterator_category<_Iter> {
 private:
+    template <input_iterator>
+    friend class basic_const_iterator;
+
     /* [[no_unique_address]] */ _Iter _Current{};
 
     using _Reference        = iter_const_reference_t<_Iter>;

--- a/tests/std/tests/P2278R4_basic_const_iterator/test.cpp
+++ b/tests/std/tests/P2278R4_basic_const_iterator/test.cpp
@@ -380,6 +380,15 @@ constexpr bool all_tests() {
     return true;
 }
 
+// GH-5321 "<xutility>: basic_const_iterator<int *> Cannot Convert to basic_const_iterator<const int *>"
+void test_conversion_instantiation() { // COMPILE-ONLY
+    struct Base {};
+    struct Derived : Base {};
+
+    [[maybe_unused]] basic_const_iterator<const int*> cit1 = basic_const_iterator<const int*>{};
+    [[maybe_unused]] basic_const_iterator<Base*> cit2      = basic_const_iterator<Derived*>{};
+}
+
 int main() {
     static_assert(all_tests());
     all_tests();

--- a/tests/std/tests/P2278R4_basic_const_iterator/test.cpp
+++ b/tests/std/tests/P2278R4_basic_const_iterator/test.cpp
@@ -374,19 +374,36 @@ constexpr void test_p2836r1() {
     }
 }
 
-constexpr bool all_tests() {
-    instantiation_test();
-    test_p2836r1();
-    return true;
-}
-
 // GH-5321 "<xutility>: basic_const_iterator<int *> Cannot Convert to basic_const_iterator<const int *>"
-void test_conversion_instantiation() { // COMPILE-ONLY
+constexpr void test_conversion_instantiation() {
     struct Base {};
     struct Derived : Base {};
 
-    [[maybe_unused]] basic_const_iterator<const int*> cit1 = basic_const_iterator<const int*>{};
-    [[maybe_unused]] basic_const_iterator<Base*> cit2      = basic_const_iterator<Derived*>{};
+    {
+        basic_const_iterator<const int*> cit = basic_const_iterator<int*>{};
+        assert(cit.base() == nullptr);
+    }
+    {
+        int n{};
+        basic_const_iterator<const int*> cit = basic_const_iterator<int*>{&n};
+        assert(cit.base() == &n);
+    }
+    {
+        basic_const_iterator<Base*> cit = basic_const_iterator<Derived*>{};
+        assert(cit.base() == nullptr);
+    }
+    {
+        Derived d{};
+        basic_const_iterator<Base*> cit = basic_const_iterator<Derived*>{&d};
+        assert(cit.base() == static_cast<Base*>(&d));
+    }
+}
+
+constexpr bool all_tests() {
+    instantiation_test();
+    test_p2836r1();
+    test_conversion_instantiation();
+    return true;
 }
 
 int main() {

